### PR TITLE
feat(atsu): ATSU as standalone system with event-bus outside of the MCDU

### PR DESCRIPTION
**Pull request by svengcz**
Fixes report about missing reset of auto-ATIS update on discord

## Summary of Changes

This PR changes the software architecture of the ATSU to an event-based system that runs as a single system outside the MCDU.

- [x] Split up ATSU into common and system
- [x] Use EventBus for the communication with the DCDU/Mailbox
- [x] Implement MCDU client to handle ATSU events outside the MCDU-page logic
- [x] Create System-Host as instrument to load Typescript-systems as single components
- [x] Remove ATSU from MCDU and add to system-host
- [x] Change flight state observer based on events
- [x] Testing, testing testing :D
  - [x] AOC
  - [x] ATC
  - [x] DCDU

## Additional context

Discord username (if different from GitHub): Sven [de en]

## Testing instructions

Test all features of the current ATSU version.
- All AOC features (ATIS, METAR, TAF, Freetext for Hoppie & FBW)
- All ATC features
  - Connect, disconnect
  - Handover
  - Message from ATC (w & w/o semantic responses (i.e. REPORT SQUAWK))
  - Message to ATC (DCL, OCL, Request, etc.)
  - Position reports (manual & automatic)
  - D-ATIS

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: src/atsu/common/src/databus/Fms.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+export interface FmsRouteData {
+    lastWaypoint: Waypoint;
+    activeWaypoint: Waypoint;
+    nextWaypoint: Waypoint;
+    destination: Waypoint;
+}
+
+export interface AtsuFmsMessages {
```
Comment (Benjozork) [inline src/atsu/common/src/databus/Fms.ts:27]: I think we should make the messages sent from ATSU -> FMS be a separate interface than the messages sent from FMS -> ATSU.

File: src/atsu/common/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import { AtsuStatusCodes } from './AtsuStatusCodes';
+import { MailboxStatusMessage } from './databus';
+import {
```
Comment (Benjozork) [inline src/atsu/common/src/index.ts:3]: You can replace this with `export * from './messages'` instead

File: src/atsu/common/src/messages/WeatherMessage.ts [hunk 3]
```diff
@@ -61,4 +61,11 @@ export class WeatherMessage extends AtsuMessage {
 
         return message;
     }
+
+    public deserialize(jsonData: Record<string, unknown>): void {
```
Comment (Benjozork) [inline src/atsu/common/src/messages/WeatherMessage.ts:65]: I would generally make those a static method instead, for example:

```ts
public static deserialize(jsonData: Record<string, unknown>): WeatherMessage
```

File: src/atsu/common/src/types/Clock.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    public hour: number;
+
+    public minute: number;
+
+    public second: number;
+
+    public secondsOfDay: number;
+
+    constructor() {
```
Comment (Benjozork) [inline src/atsu/common/src/types/Clock.ts:16]: You can make a shorthand for this by declaring properties in the constructor:

```ts
constructor(
    public year: number,
    public month: number,
    public dayOfMonth: number,
    public hour: number,
    public minute: number,
    public second: number,
    public secondsOfDay: number,
) {
}
```

TypeScript will automatically generate the assignments in the constructor body at compile-time. The properties in the class body can then be omitted.

File: src/atsu/common/src/types/Waypoint.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+export class Waypoint {
```
Comment (Benjozork) [inline src/atsu/common/src/types/Waypoint.ts:1]: This looks like it could just be an interface.

File: src/atsu/system/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { Atsu } from './ATSU';
```
Comment (Benjozork) [inline src/atsu/system/src/index.ts:1]: We can replace those with `export * from './X'`

File: src/instruments/src/DCDU/index.tsx [hunk 7]
```diff
@@ -232,140 +237,132 @@ const DCDU: React.FC = () => {
...
                 }
+                setMessages(new Map<number, DcduMessageBlock>(messagesRef.current));
             }
-            setMessages(new Map<number, DcduMessageBlock>(messagesRef.current));
-        }
-    });
+        });
+
+        setSubscriber(newSubscriber);
+    }, []);
```
Comment (Benjozork) [inline src/instruments/src/DCDU/index.tsx:365]: We need a way here to unsubscribe from all those things in the cleanup callback, but unfortunately our version of `msfssdk` does not have the `Subscription` type....

Need to think of a way to handle this, or add a TODO. This component will never unmount for now, but it is still unoptimal to leak subscriptions

**Review by Benjozork — CHANGES_REQUESTED** [review]

File: .github/CHANGELOG.md [hunk 8]
```diff
@@ -19,8 +19,10 @@
 1. [EFB] Fix and improve pushback system and add API documentation - @frankkopp (Frank Kopp)
 1. [RMP] RMPs navigation backup - Julian Sebline (Julian Sebline#8476 on Discord)
 1. [SEC] Fix GND SPLR logic, add missing GND SPLR partial extension condition - @lukecologne (luke)
-1. [FMGC] Improved importing flight plans from MSFS World Map - @frankkopp (Frank Kopp)  
-1. [EFB] Added boarding time indication to Payload page - @ChristianLutzCL (Christian Lutz) @frankkopp (Frank Kopp)
+1. [FMGC] Improved importing flight plans from MSFS World Map - @frankkopp (Frank Kopp)
+1. [EFB] Added boarding time indication to Payload page - @ChristianLutzCL (Christian Lutz) @frankkopp(Frank Kopp)
```
[the remark this answers is not in the document]
Comment (svengcz) [reply .github/CHANGELOG.md:23]: No, reverted the commit

File: src/atsu/common/src/index.ts (hunk 2 above)
Comment (svengcz) [reply src/atsu/common/src/index.ts:3]: fixed

File: src/atsu/system/src/index.ts (hunk 6 above)
Comment (svengcz) [reply src/atsu/system/src/index.ts:1]: fixed

File: src/atsu/common/src/types/Waypoint.ts (hunk 5 above)
Comment (svengcz) [reply src/atsu/common/src/types/Waypoint.ts:1]: fixed

File: src/instruments/src/DCDU/index.tsx (hunk 7 above)
Comment (svengcz) [reply src/instruments/src/DCDU/index.tsx:365]: handled the leak in the cleanup-call of the effect

File: src/atsu/common/src/messages/WeatherMessage.ts (hunk 3 above)
Comment (svengcz) [reply src/atsu/common/src/messages/WeatherMessage.ts:65]: fixed

File: src/atsu/common/src/databus/Fms.ts (hunk 1 above)
Comment (svengcz) [reply src/atsu/common/src/databus/Fms.ts:27]: fixed

**Review by Benjozork — APPROVED** [review]

**Comment by frankkopp** [comment]
I'd really love to see some documentation esp. around the systems-host - this is really a great addition and it should be clearly documented when to use it, how to use it and how additional modules should be structured. 

Either a README in the systems-host folder or directly in the index.tsx would work fine.

This way others could learn faster how this should be used in a way you intended it to be used. Maybe a minimal example would make sense - Atsu is a big system and it might be easier to understand a more "hello world" like example on systems-host usage.

**Pull request merged**